starlette/middleware/limits.py

```diff
@@ -0,0 +1,66 @@
+"""Middleware that limits the body size of incoming requests."""
+from starlette.datastructures import Headers
+from starlette.responses import PlainTextResponse
+from starlette.types import ASGIApp, Message, Receive, Scope, Send
+
+DEFAULT_MAX_BODY_SIZE = 2_621_440  # 2.5MB
+
+
+class ContentTooLarge(Exception):
+    def __init__(self, max_body_size: int) -> None:
+        self.max_body_size = max_body_size
```

Comment:
This shouldn't be an `HTTPException`. The `user_middleware`s run after the `ExceptionMiddleware`.

But also, I think it should be just this, as we do in other parts of the code.

Reply:
Are you sure? There’s currently two places where exceptions are caught. Once in the route and once at the app middleware level. I think we’ll remove the latter eventually.

Reply:
Right. What I said doesn't make sense.

I still don't want to use the `HTTPException`, because we'll need to check if `app` is in the scope in any case.

Reply:
> I still don't want to use the HTTPException, because we'll need to check if app is in the scope in any case.

Why? My idea was that one of two things would happen:
- `BodySizeLimitMiddleware -> ExceptionMiddleware -> App`: `App` raises a `ContentTooLarge` exception. That gets caught by `ExceptionMiddleware` which returns the 413 status code and message. `BodySizeLimitMiddleware` does nothing and never catches `ContentTooLarge`. This would be the scenario where you wrap an individual route.
- `BodySizeLimitMiddleware -> SomeOtherMiddleware -> ExceptionMiddleware -> App`. Where `SomeOtherMiddleware` might read the body (e.g. `BaseHTTPMiddleware`). In this case, there may be no `ExceptionMiddleware` to catch the exception (currently there is but as per above we may remove it in the future, also we shouldn't make assumptions about other middleware) so `BodySizeLimitMiddleware` itself catches the exception and returns the 413 response. This is the scenario where an entire app or router is wrapped, or the middleware is being used outside of a Starlette context.

Reply:
I was wrong on my last message. It does need the `if 'app' in scope:`.

Reply:
I'll push the code to show you.

Reply:
This is the commit: https://github.com/encode/starlette/pull/2350/commits/c4e7aad4c5beae118890348af0fcb9d434ef7368

I'll remove the `HTTPException` inheritance.

Reply:
Look as long as it works both with and without `ExceptionMiddleware` I'm happy